### PR TITLE
Use demangled name for BuiltinWithGenericPointer

### PR DESCRIPTION
**Pull request by bmanga**
This tries to address #1343 

@rjodinchr should I also add a test?

### CODE REVIEW

**Comment by rjodinchr** [comment]
> should I also add a test?

It would be great to add the simple example you gave in the issue. Maybe just running clang front-end as in https://github.com/google/clspv/blob/main/test/emit_ir.cl#L1

File: lib/Builtins.cpp [hunk 1]
```diff
@@ -777,11 +778,32 @@ glsl::ExtInst Builtins::getDirectOrIndirectExtInstEnum(
...
 bool Builtins::BuiltinWithGenericPointer(StringRef name) {
-  if (name.contains("fract") || name.contains("frexp") ||
-      name.contains("modf") || name.contains("remquo") ||
-      name.contains("lgamma_r") || name.contains("vstore_half") ||
-      name.contains("sincos"))
-    return true;
+  if (!name.starts_with("_Z")) {
+    return false;
+  }
+  size_t pos = 2;
```
Comment (rjodinchr) [inline lib/Builtins.cpp:784]:
```suggestion
  StringRef prefix("_Z");
  if (!name.starts_with(prefix)) {
    return false;
  }
  size_t pos = prefix.size();
```

File: lib/Builtins.cpp [hunk 2]
```diff
@@ -777,11 +778,32 @@ glsl::ExtInst Builtins::getDirectOrIndirectExtInstEnum(
...
+                        "remquo", "lgamma_r", "sincos"};
+
+  for (const auto &fn_name : list) {
+    if (demangled == fn_name) {
+      return true;
+    }
+  }
+
+  // Check for vstore_half family of functions.
+  if (demangled.size() > 10) {
```
Comment (rjodinchr) [inline lib/Builtins.cpp:801]:
```suggestion
  if (demangled.size() >= strlen("vstore_halfx_rtx") {
```

File: test/BuiltinsWithGenericPointer/issue-1343.cl [hunk 3]
```diff
@@ -0,0 +1,5 @@
+// RUN: clspv --cl-std=CLC++ --inline-entry-points %s -o %t.spv
```
Comment (rjodinchr) [inline test/BuiltinsWithGenericPointer/issue-1343.cl:1]:
```suggestion
// RUN: clspv --cl-std=CLC++ --inline-entry-points --output-format=ll %s -o %t.ll
```

File: test/BuiltinsWithGenericPointer/issue-1343.cl [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+
```
Comment (rjodinchr) [inline test/BuiltinsWithGenericPointer/issue-1343.cl:2]: Please add some checks:
```
// RUN: FileCheck %s < %t.ll
// CHECK: something
```

**Review by rjodinchr — CHANGES_REQUESTED** [review]

**Comment by rjodinchr** [comment]
We don't like adding a test without explicit checks. It makes it quite complicated sometimes when it starts failing for whatever reason to know what the test was expecting to check.

As it's the link that is failing and it is performed after the GenerateIRFile, it makes sense that it does not work with `--output-format`.
What we can do instead is to use `--print-before=annotation-to-metadata` to get the `ll` file after the link. So that we can check that the content is what we expect and nothing more.

**Comment by bmanga** [comment]
Mh I've tried that and it seems like clspv crashes before being able to dump anything to the file...

**Comment by rjodinchr** [comment]
Without the fix, it should crash before anything is dumped. Am I missing something?

**Comment by bmanga** [comment]
Ah ok I misunderstood your previous comment. I thought you wanted the test to fail through FileCheck and I assumed that flag would output something I could act upon without crashing the compiler.

Updated. Let me know if that's ok :)

File: test/BuiltinsWithGenericPointer/issue-1343.cl [hunk 5]
```diff
@@ -0,0 +1,12 @@
+// RUN: clspv --cl-std=CLC++ --inline-entry-points %s --print-before=annotation-to-metadata --output-format=ll -O3 -o %t.ll
+// RUN: FileCheck %s < %t.ll
+
+// CHECK: define dso_local spir_func float @_Z22contains_frexp_in_namefPU3AS4i(float %0, ptr addrspace(4) %1)
+// CHECK: define dso_local spir_func void @_Z12vstore_half5Dv2_fjPU3AS4Dh(<2 x float> %0, i32 %1, ptr addrspace(4) %2)
+// CHECK: define dso_local spir_func void @_Z13xvstore_half2Dv2_fjPU3AS4Dh(<2 x float> %0, i32 %1, ptr addrspace(4) %2)
+
```
Comment (rjodinchr) [inline test/BuiltinsWithGenericPointer/issue-1343.cl:7]: The way it is right now is not failing if I do not use the version of `BuiltinWithGenericPointer` in this CL.
```suggestion
// RUN: clspv --cl-std=CLC++ --inline-entry-points %s --print-before=annotation-to-metadata -o %t.spv &> %t.ll
// RUN: FileCheck %s < %t.ll
// RUN: spirv-val %t.spv

// CHECK: define dso_local spir_func float @_Z22contains_frexp_in_namefPU3AS4i(float %0, ptr addrspace(4) %1)
// CHECK: define dso_local spir_func void @_Z12vstore_half5Dv2_fjPU3AS4Dh(<2 x float> %0, i32 %1, ptr addrspace(4) %2)
// CHECK: define dso_local spir_func void @_Z13xvstore_half2Dv2_fjPU3AS4Dh(<2 x float> %0, i32 %1, ptr addrspace(4) %2)

void kernel foo() {}
```

**Review by rjodinchr — CHANGES_REQUESTED** [review]

File: test/BuiltinsWithGenericPointer/issue-1343.cl (hunk 5 above)
Comment (bmanga) [reply test/BuiltinsWithGenericPointer/issue-1343.cl:7]: Ah ok... I'm pretty sure I tested it but maybe I got something wrong. I'll commit your suggestion :)

**Pull request merged**